File: java/spring-boot-mongodb/src/test/java/com/meridal/examples/springbootmongo/domain/RecordingTest.java
package com.meridal.examples.springbootmongo.domain;

import com.meridal.examples.springbootmongo.test.TestFramework;
import org.junit.jupiter.api.Test;

import java.util.ArrayList;
import java.util.List;

import static org.junit.jupiter.api.Assertions.*;


public class RecordingTest extends TestFramework {

    private static final String ID = "1234";
    private static final String OTHER_ID = "5678";

    @Test
    public void testEqualsAndHashCodeWithSameID() {
        final Recording recording = this.randomRecording(ID);
        final Recording other = this.cloneRecording(ID, recording);
        assertEquals(recording, other);
        assertEquals(recording.hashCode(), other.hashCode());
    }

    @Test
    public void testEqualsAndHashCodeWithDifferentID() {
        final Recording recording = this.randomRecording(ID);
        final Recording other = this.cloneRecording(OTHER_ID, recording);
        assertNotEquals(recording, other);
        assertNotEquals(recording.hashCode(), other.hashCode());
    }

    @Test
    public void testEqualsAndHashCodeWithNullID() {
        final Recording recording = this.randomRecording();
        final Recording other = this.cloneRecording(ID, recording);
        assertNotEquals(recording, other);
        assertNotEquals(recording.hashCode(), other.hashCode());
    }

    @Test
    public void testEqualsAndHashCodeWithBothIDsNull() {
        final Recording recording = this.randomRecording();
        final Recording other = this.cloneRecording(null, recording);
        assertNotEquals(recording, other);
        assertEquals(recording.hashCode(), other.hashCode());
    }

    @Test
    public void testEqualsWithAnotherType() {
        final Recording recording = this.randomRecording();
        final List<Integer> other = new ArrayList<>();
        assertFalse(recording.equals(other));
   }

    @Test
    public void testEqualsWithNull() {
        final Recording recording = this.randomRecording();
        assertFalse(recording.equals(null));
    }
}

File: kotlin/kotlin-spring-boot-mysql/src/test/kotlin/com/meridal/examples/kotlinspringboot/model/RecordingTest.kt
package com.meridal.examples.kotlinspringboot.model

import org.junit.jupiter.api.Assertions.*

import com.meridal.examples.kotlinspringboot.test.TestFramework
import org.junit.jupiter.api.Test

class RecordingTest(): TestFramework() {

    private val ID = 1234L
    private val OTHER_ID = 5678L
    private val NOT_PERSISTED = 0L;

    @Test
    fun `Recordings with the same ID are equal and have the same hashCode`() {
        val recording = randomRecording(ID)
        val other = cloneRecording(ID, recording)
        assertEquals(recording, other)
        assertEquals(recording.hashCode(), other.hashCode())
    }

    @Test
    fun `Recordings with different IDs are not equal and do not have the same hashCode`() {
        val recording = randomRecording(ID)
        val other = cloneRecording(OTHER_ID, recording)
        assertNotEquals(recording, other)
        assertNotEquals(recording.hashCode(), other.hashCode())
    }

    @Test
    fun `Recordings that have not been persisted are not equal`() {
        val recording = randomRecording(ID)
        val other = cloneRecording(NOT_PERSISTED, recording)
        val another = cloneRecording(NOT_PERSISTED, recording)
        assertNotEquals(recording, other)
        assertNotEquals(other, another)
    }

    @Test
    fun `Recordings are not equal to other types`() {
        val recording = randomRecording(ID)
        val list = listOf(1, 2, 3)
        assertFalse(recording == list)
    }

    @Test
    fun `A recording is not equal to a null object`() {
        val recording = randomRecording(ID)
        val other: Recording? = null
        assertNotEquals(recording, other)
    }
}